File: backend/app/schemas/exam_scheduling.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ExamModel(BaseModel):
    model_config = ConfigDict(alias_generator=lambda value: value.split("_")[0] + "".join(part.title() for part in value.split("_")[1:]), populate_by_name=True, extra="forbid")


class OperationError(ExamModel):
    code: str
    message: str
    field: str | None = None


class ErrorResponse(ExamModel):
    errors: list[OperationError]


class ExamConfigurationInput(ExamModel):
    identifier: str
    duration_minutes: int
    recommended_start_override: date | None
    recommended_end_override: date | None
    required_capacity: int
    exam_type: str
    responsible_lecturer_id: int


class SaveExamConfigurationRequest(ExamModel):
    semester_id: int = Field(ge=1)
    enabled: bool
    expected_revision: int | None = Field(ge=1)
    configuration: ExamConfigurationInput | None


class ExamConfiguration(ExamModel):
    id: int
    revision: int
    identifier: str
    duration_minutes: int
    recommended_start_override: date | None
    recommended_end_override: date | None
    required_capacity: int
    exam_type: str
    responsible_lecturer_id: int
    configuration_consumed: bool
    recommended_start_date: date | None
    recommended_end_date: date | None
    recommendation_was_overridden: bool


class FinalTeachingAnchor(ExamModel):
    date: date
    end_time: str
    teaching_session_id: int


class GenerationEligibility(ExamModel):
    eligible: bool
    code: str | None
    message: str | None


class ResourceReference(ExamModel):
    id: int
    name: str
    reference_code: str | None = None


class RoomReference(ResourceReference):
    capacity: int


class ExamIssue(ExamModel):
    code: str
    message: str
    related_date: date | None = None
    related_resource: ResourceReference | None = None
    related_session_id: int | None = None
    holiday_name: str | None = None


class ExamSessionResponse(ExamModel):
    id: int
    revision: int
    course_id: int
    semester_id: int
    configuration_identifier: str
    exam_type: str
    duration_minutes: int
    required_capacity: int
    recommended_start_date: date
    recommended_end_date: date
    recommendation_was_overridden: bool
    outside_recommended_window: bool
    final_teaching_anchor: FinalTeachingAnchor
    date: date
    start_time: str
    end_time: str
    lecturer: ResourceReference
    cohort: ResourceReference
    room: RoomReference
    lifecycle_status: Literal["active", "past"]
    source: Literal["generated", "manual"]
    validity_issues: list[ExamIssue]
    input_snapshot_token: str


class ExamCoursePlanningState(ExamModel):
    course_id: int
    course_name: str
    semester_id: int
    cohort_id: int
    cohort_name: str
    enabled: bool
    configuration: ExamConfiguration | None
    final_teaching_anchor: FinalTeachingAnchor | None
    active_exam: ExamSessionResponse | None
    past_exams: list[ExamSessionResponse]
    generation_eligibility: GenerationEligibility
    input_snapshot_token: str


class ExamPlanningOverview(ExamModel):
    semester_id: int
    institution_today: date
    courses: list[ExamCoursePlanningState]


class ExamGenerationPreparationRequest(ExamModel):
    semester_id: int = Field(ge=1)
    schedule_revision_id: int = Field(ge=1)
    course_ids: list[int] = Field(min_length=1, max_length=100)

    @model_validator(mode="after")
    def unique_courses(self):
        if len(set(self.course_ids)) != len(self.course_ids) or any(value < 1 for value in self.course_ids):
            raise ValueError("courseIds must be unique.")
        return self


class PreparedExamCourse(ExamModel):
    course_id: int
    course_name: str
    configuration_id: int | None
    configuration_revision: int | None
    input_snapshot_token: str
    eligibility: GenerationEligibility


class ExamGenerationPreparation(ExamModel):
    semester_id: int
    schedule_revision_id: int | None
    institution_today: date
    shared_snapshot_token: str
    courses: list[PreparedExamCourse]


class PreparedExamCourseInput(ExamModel):
    course_id: int = Field(ge=1)
    configuration_id: int | None
    configuration_revision: int | None
    input_snapshot_token: str


class GenerateExamsRequest(ExamModel):
    semester_id: int = Field(ge=1)
    schedule_revision_id: int = Field(ge=1)
    institution_today: date
    shared_snapshot_token: str
    courses: list[PreparedExamCourseInput] = Field(min_length=1, max_length=100)

    @model_validator(mode="after")
    def unique_courses(self):
        if len({item.course_id for item in self.courses}) != len(self.courses):
            raise ValueError("courses must contain unique courseIds.")
        return self


class ExamGenerationOutcome(ExamModel):
    course_id: int
    course_name: str
    configuration_id: int | None
    configuration_identifier: str | None
    status: Literal["scheduled", "failed", "stale", "skipped_active", "skipped_disabled"]
    saved: bool
    exam: ExamSessionResponse | None
    reasons: list[ExamIssue]


class ExamGenerationSummary(ExamModel):
    total: int
    scheduled: int
    failed: int
    stale: int
    skipped_active: int
    skipped_disabled: int
    elapsed_milliseconds: int
    optimal_for_prepared_snapshot: bool


class ExamGenerationResult(ExamModel):
    semester_id: int
    summary: ExamGenerationSummary
    outcomes: list[ExamGenerationOutcome]


class CreateManualExamRequest(ExamModel):
    semester_id: int = Field(ge=1)
    schedule_revision_id: int = Field(ge=1)
    date: date
    start_time: str = Field(pattern=r"^([01]\d|2[0-3]):[0-5]\d$")
    lecturer_id: int = Field(ge=1)
    room_id: int = Field(ge=1)
    expected_configuration_revision: int = Field(ge=1)
    input_snapshot_token: str


class UpdateExamRequest(ExamModel):
    schedule_revision_id: int = Field(ge=1)
    date: date
    start_time: str = Field(pattern=r"^([01]\d|2[0-3]):[0-5]\d$")
    lecturer_id: int = Field(ge=1)
    room_id: int = Field(ge=1)
    expected_exam_revision: int = Field(ge=1)
    input_snapshot_token: str


class DeleteExamRequest(ExamModel):
    schedule_revision_id: int = Field(ge=1)
    confirmed: Literal[True]
    expected_exam_revision: int = Field(ge=1)
    input_snapshot_token: str


class DeleteExamResponse(ExamModel):
    deleted_exam_id: int
    deleted_lifecycle_status: Literal["active", "past"]
    consequence: Literal["configuration_enabled_unscheduled", "historical_exam_only"]
    state: ExamCoursePlanningState

File: client/src/api/examScheduling.ts
export type ExamIssue = { code: string; message: string; relatedDate: string | null; relatedResource: ResourceReference | null; relatedSessionId: number | null; holidayName: string | null }
export type ResourceReference = { id: number; name: string; referenceCode: string | null }
export type RoomReference = ResourceReference & { capacity: number }
export type FinalTeachingAnchor = { date: string; endTime: string; teachingSessionId: number }
export type GenerationEligibility = { eligible: boolean; code: 'ELIGIBLE' | 'DISABLED' | 'CONFIGURATION_INCOMPLETE' | 'FINAL_TEACHING_SESSION_MISSING' | 'AUTOMATIC_START_TIME_UNAVAILABLE' | 'ACTIVE_EXAM_EXISTS' | 'CONFIGURATION_CONSUMED' | null; message: string | null }
export type ExamConfigurationInput = { identifier: string; durationMinutes: number; recommendedStartOverride: string | null; recommendedEndOverride: string | null; requiredCapacity: number; examType: string; responsibleLecturerId: number }
export type ExamConfiguration = ExamConfigurationInput & { id: number; revision: number; configurationConsumed: boolean; recommendedStartDate: string | null; recommendedEndDate: string | null; recommendationWasOverridden: boolean }
export type ExamSession = { id: number; revision: number; courseId: number; semesterId: number; configurationIdentifier: string; examType: string; durationMinutes: number; requiredCapacity: number; recommendedStartDate: string; recommendedEndDate: string; recommendationWasOverridden: boolean; outsideRecommendedWindow: boolean; finalTeachingAnchor: FinalTeachingAnchor; date: string; startTime: string; endTime: string; lecturer: ResourceReference; cohort: ResourceReference; room: RoomReference; lifecycleStatus: 'active' | 'past'; source: 'generated' | 'manual'; validityIssues: ExamIssue[]; inputSnapshotToken: string }
export type ExamCoursePlanningState = { courseId: number; courseName: string; semesterId: number; cohortId: number; cohortName: string; enabled: boolean; configuration: ExamConfiguration | null; finalTeachingAnchor: FinalTeachingAnchor | null; activeExam: ExamSession | null; pastExams: ExamSession[]; generationEligibility: GenerationEligibility; inputSnapshotToken: string }
export type ExamPlanningOverview = { semesterId: number; institutionToday: string; courses: ExamCoursePlanningState[] }
export type SaveExamConfigurationRequest = { semesterId: number; enabled: boolean; expectedRevision: number | null; configuration: ExamConfigurationInput | null }
export type PreparedExamCourse = { courseId: number; courseName: string; configurationId: number | null; configurationRevision: number | null; inputSnapshotToken: string; eligibility: GenerationEligibility }
export type ExamGenerationPreparation = { semesterId: number; scheduleRevisionId: number; institutionToday: string; sharedSnapshotToken: string; courses: PreparedExamCourse[] }
export type PreparedExamCourseInput = Omit<PreparedExamCourse, 'courseName' | 'eligibility'>
export type GenerateExamsRequest = { semesterId: number; scheduleRevisionId: number; institutionToday: string; sharedSnapshotToken: string; courses: PreparedExamCourseInput[] }
export type ExamGenerationOutcome = { courseId: number; courseName: string; configurationId: number | null; configurationIdentifier: string | null; status: 'scheduled' | 'failed' | 'stale' | 'skipped_active' | 'skipped_disabled'; saved: boolean; exam: ExamSession | null; reasons: ExamIssue[] }
export type ExamGenerationResult = { semesterId: number; summary: { total: number; scheduled: number; failed: number; stale: number; skippedActive: number; skippedDisabled: number; elapsedMilliseconds: number; optimalForPreparedSnapshot: boolean }; outcomes: ExamGenerationOutcome[] }
export type CreateManualExamRequest = { semesterId: number; scheduleRevisionId: number; date: string; startTime: string; lecturerId: number; roomId: number; expectedConfigurationRevision: number; inputSnapshotToken: string }
export type UpdateExamRequest = { scheduleRevisionId: number; date: string; startTime: string; lecturerId: number; roomId: number; expectedExamRevision: number; inputSnapshotToken: string }
export type DeleteExamRequest = { scheduleRevisionId: number; confirmed: true; expectedExamRevision: number; inputSnapshotToken: string }
export type DeleteExamResponse = { deletedExamId: number; deletedLifecycleStatus: 'active' | 'past'; consequence: 'configuration_enabled_unscheduled' | 'historical_exam_only'; state: ExamCoursePlanningState }
export type ExamOperationError = { code: string; message: string; field: string | null; meta?: Record<string, unknown> | null }

export class ExamSchedulingApiError extends Error {
  status: number
  errors: ExamOperationError[]
  currentState: ExamCoursePlanningState | null
  constructor(status: number, errors: ExamOperationError[], currentState: ExamCoursePlanningState | null = null) {
    super(errors[0]?.message ?? 'Exam scheduling request failed.')
    this.status = status; this.errors = errors; this.currentState = currentState
  }
}

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? ''
const json = (method: string, body: unknown): RequestInit => ({ method, headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(body) })

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${path}`, init)
  const payload = await response.json().catch(() => null)
  if (!response.ok) throw new ExamSchedulingApiError(response.status, payload?.errors ?? [{ code: 'REQUEST_FAILED', message: 'Exam scheduling request failed.', field: null }], payload?.currentState ?? null)
  return payload as T
}

export const getExamPlanningOverview = (semesterId: number) => request<ExamPlanningOverview>(`/api/exam-planning?semesterId=${semesterId}`)
export const saveExamConfiguration = (courseId: number, input: SaveExamConfigurationRequest) => request<ExamCoursePlanningState>(`/api/courses/${courseId}/exam-configuration`, json('PUT', input))
export function prepareExamGeneration(semesterId: number, scheduleRevisionId: number, courseIds: number[]) {
  if (courseIds.length < 1 || courseIds.length > 100 || new Set(courseIds).size !== courseIds.length) return Promise.reject(new ExamSchedulingApiError(422, [{ code: 'INVALID_SELECTION', message: 'Select 1 to 100 unique courses.', field: 'courseIds' }]))
  return request<ExamGenerationPreparation>('/api/exams/generation/prepare', json('POST', { semesterId, scheduleRevisionId, courseIds }))
}
export const generateExams = (input: GenerateExamsRequest) => request<ExamGenerationResult>('/api/exams/generation', json('POST', input))
export const createManualExam = (courseId: number, input: CreateManualExamRequest) => request<ExamCoursePlanningState>(`/api/courses/${courseId}/exam-sessions`, json('POST', input))
export const updateExam = (examId: number, input: UpdateExamRequest) => request<ExamCoursePlanningState>(`/api/exam-sessions/${examId}`, json('PATCH', input))
export const deleteExam = (examId: number, input: DeleteExamRequest) => request<DeleteExamResponse>(`/api/exam-sessions/${examId}`, json('DELETE', input))
